SteamKit2/SteamKit2/Util/HardwareUtils.cs

```diff
@@ -0,0 +1,353 @@
+                }
+            }
+
+            return base.GetMachineGuid();
+        }
+
+        public override byte[] GetDiskId()
+        {
+            var stat = new statfs();
+            var statted = statfs64( "/", ref stat );
+            if ( statted == 0 )
+            {
+                using ( var session = DASessionCreate( CFTypeRef.None ) )
+                using ( var disk = DADiskCreateFromBSDName( CFTypeRef.None, session, stat.f_mntfromname ) )
+                using ( var properties = DADiskCopyDescription( disk ) )
+                using ( var key = CFStringCreateWithCString( CFTypeRef.None, DiskArbitration.kDADiskDescriptionMediaUUIDKey, CFStringEncoding.kCFStringEncodingASCII ) )
+                {
+                    IntPtr cfuuid = IntPtr.Zero;
+                    if ( CFDictionaryGetValueIfPresent( properties, key, out cfuuid ) )
+                    {
+                        using ( var uuidString = CFUUIDCreateString( CFTypeRef.None, cfuuid ) )
+                        {
+                            var stringBuilder = new StringBuilder( 64 );
+                            if ( CFStringGetCString( uuidString, stringBuilder, stringBuilder.Capacity, CFStringEncoding.kCFStringEncodingASCII ) )
+                            {
+                                return Encoding.ASCII.GetBytes( stringBuilder.ToString() );
+                            }
+                        }
+                    }
+                }
+            }
+
+            return base.GetDiskId();
+        }
+    }
+
+    static class HardwareUtils
+    {
+        class MachineID : MessageObject
+        {
+            public MachineID()
+                : base()
+            {
+                this.KeyValues["BB3"] = new KeyValue();
+                this.KeyValues["FF2"] = new KeyValue();
+                this.KeyValues["3B3"] = new KeyValue();
+            }
+
+
+            public void SetBB3( string value )
+            {
+                this.KeyValues["BB3"].Value = value;
+            }
+
+            public void SetFF2( string value )
+            {
+                this.KeyValues["FF2"].Value = value;
+            }
+
+            public void Set3B3( string value )
+            {
+                this.KeyValues["3B3"].Value = value;
+            }
+
+            public void Set333( string value )
+            {
+                this.KeyValues["333"] = new KeyValue( value: value );
+            }
+        }
+
+        public static byte[] GenerateMachineID()
+        {
+            // the aug 25th 2015 CM update made well-formed machine MessageObjects required for logon
+            // this was flipped off shortly after the update rolled out, likely due to linux steamclients running on distros without a way to build a machineid
+            // so while a valid MO isn't currently (as of aug 25th) required, they could be in the future and we'll abide by The Valve Law now
+
+            var machineId = new MachineID();
```

Comment:
Should this function cache constructed machine id, and just return the cached object on future calls? WMI is slow, and the result isn't going to cache between calls.